activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -406,6 +449,43 @@ def reap
       end
 
       private
+      def with_exclusively_acquired_all_connections
+        with_new_connections_blocked do
+          checkout_all_existing_connections!
+          yield
+        end
+      end
+
+      def checkout_all_existing_connections!
+        collected_conns = synchronize do
+          # account for our own connections
+          @connections.select {|conn| conn.owner == Thread.current}
+        end
+
+        newly_checked_out = []
+        while true
+          synchronize do
+            return if collected_conns.size == @connections.size && @now_connecting == 0
+            collected_conns << (conn = checkout)
+            newly_checked_out << conn
+          end
+        end
+      rescue Exception # if something went wrong, release what was gathered and re-raise
+        # releasing only those conns that were checked out in this method, conns
+        # checked outside this method (before it was called) are not for us to release
+        newly_checked_out.each {|conn| checkin(conn)}
+        raise
+      end
+
+      def with_new_connections_blocked
+        previous_value = nil
+        synchronize do
+          previous_value, @new_cons_enabled = @new_cons_enabled, false
+        end
+        yield
+      ensure
+        synchronize { @new_cons_enabled = previous_value }
```

Comment:
Looks like that `with_new_connections_blocked` have a race condition. This method sometimes causes `@new_cons_enabled = false` then never toggled to true.